File: research-workflow-agent/resilience.py
"""Small, observable fallback layer for models and external research services."""

from __future__ import annotations

import hashlib
import json
import threading
from dataclasses import asdict, dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Callable


def operation_id(service: str, payload: Any) -> str:
    encoded = json.dumps(payload, ensure_ascii=False, sort_keys=True, default=str).encode("utf-8")
    return f"{service}_{hashlib.sha256(encoded).hexdigest()[:20]}"


def classify_complexity(text: str) -> str:
    """A deterministic routing hint; it never delegates authorization to a model."""
    complex_markers = ("比较", "综合", "评审", "架构", "多篇", "experiment",
                       "compare", "synthesize", "review", "architecture")
    return "strong" if len(text) > 180 or any(marker in text.lower() for marker in complex_markers) else "light"


@dataclass(frozen=True)
class ModelEndpoint:
    name: str
    client: Any
    model: str
    tier: str = "strong"
    primary: bool = False


class FallbackJournal:
    """Append-only switch journal plus idempotent results and deferred operations."""

    def __init__(self, data_dir: Path):
        self.events_path = data_dir / "fallback_events.jsonl"
        self.cache_path = data_dir / "fallback_results.json"
        self.pending_path = data_dir / "pending_external_tasks.json"
        self._lock = threading.RLock()

    def execute(self, service: str, op_id: str, providers: list[tuple[str, Callable]],
                cache_result: bool = False) -> Any:
        with self._lock:
            cached = self._read(self.cache_path, {}).get(op_id)
        if cached is not None:
            self.log(op_id, service, "cache", "cache", "reused", "idempotent replay")
            return cached
        errors = []
        for index, (name, operation) in enumerate(providers):
            try:
                result = operation()
                previous = providers[index - 1][0] if index else name
                self.log(op_id, service, previous, name, "succeeded",
                         "primary" if index == 0 else "fallback")
                if cache_result:
                    with self._lock:
                        cache = self._read(self.cache_path, {})
                        cache[op_id] = result
                        self._write(self.cache_path, cache)
                return result
            except Exception as exc:
                errors.append(f"{name}: {exc}")
                target = providers[index + 1][0] if index + 1 < len(providers) else "pending"
                self.log(op_id, service, name, target, "failed", str(exc))
        raise RuntimeError("; ".join(errors))

    def defer(self, service: str, payload: dict[str, Any], reason: str) -> str:
        op_id = operation_id(service, payload)
        with self._lock:
            pending = self._read(self.pending_path, {})
            if op_id not in pending:
                pending[op_id] = {"id": op_id, "service": service, "payload": payload,
                                  "reason": reason, "status": "pending", "created_at": self._now()}
                self._write(self.pending_path, pending)
        self.log(op_id, service, "all", "pending", "deferred", reason)
        return op_id

    def log(self, op_id: str, service: str, source: str, target: str,
            status: str, reason: str) -> None:
        record = {"time": self._now(), "operation_id": op_id, "service": service,
                  "from": source, "to": target, "status": status, "reason": reason}
        with self._lock:
            self.events_path.parent.mkdir(parents=True, exist_ok=True)
            with self.events_path.open("a", encoding="utf-8") as handle:
                handle.write(json.dumps(record, ensure_ascii=False) + "\n")

    @staticmethod
    def _now() -> str:
        return datetime.now(timezone.utc).isoformat(timespec="seconds")

    @staticmethod
    def _read(path: Path, default: Any) -> Any:
        return json.loads(path.read_text(encoding="utf-8")) if path.exists() else default

    @staticmethod
    def _write(path: Path, value: Any) -> None:
        path.parent.mkdir(parents=True, exist_ok=True)
        temporary = path.with_suffix(path.suffix + ".tmp")
        temporary.write_text(json.dumps(value, ensure_ascii=False, indent=2), encoding="utf-8")
        temporary.replace(path)


class ResilientModelClient:
    """Anthropic-compatible client that routes by complexity and fails over."""

    def __init__(self, endpoints: list[ModelEndpoint], journal: FallbackJournal):
        if not endpoints:
            raise ValueError("At least one model endpoint is required")
        self.endpoints, self.journal = endpoints, journal
        self.messages = self

    def _ordered(self, complexity: str) -> list[ModelEndpoint]:
        matching = [item for item in self.endpoints if item.tier == complexity]
        remaining = [item for item in self.endpoints if item not in matching]
        return sorted(matching, key=lambda item: not item.primary) + sorted(
            remaining, key=lambda item: not item.primary)

    def create_for_task(self, complexity: str, **kwargs):
        endpoints = self._ordered(complexity)
        op_id = operation_id("model", {"complexity": complexity,
                                        "messages": kwargs.get("messages", [])})
        providers = []
        for endpoint in endpoints:
            request = dict(kwargs)
            request["model"] = endpoint.model
            providers.append((endpoint.name,
                              lambda endpoint=endpoint, request=request:
                              endpoint.client.messages.create(**request)))
        return self.journal.execute("model", op_id, providers)

    def create(self, **kwargs):
        return self.create_for_task("strong", **kwargs)


class FallbackEmbeddingModel:
    """Use a local embedding when the configured remote embedding fails."""

    def __init__(self, primary: Any, fallback: Any, journal: FallbackJournal):
        self.primary, self.fallback, self.journal = primary, fallback, journal
        self.dimensions = getattr(primary, "dimensions", getattr(fallback, "dimensions", 0))

    def embed_many(self, texts: list[str]) -> list[list[float]]:
        op_id = operation_id("embedding", texts)
        return self.journal.execute("embedding", op_id, [
            ("remote", lambda: self.primary.embed_many(texts)),
            ("local", lambda: self.fallback.embed_many(texts)),
        ])

    def embed(self, text: str) -> list[float]:
        return self.embed_many([text])[0]


class FallbackKnowledgeBase:
    """Mirror writes locally and use the local index when Elasticsearch fails."""

    def __init__(self, primary: Any, fallback: Any, journal: FallbackJournal):
        self.primary, self.fallback, self.journal = primary, fallback, journal

    def add_document(self, title: str, content: str, source: str = "manual",
                     chunk_size: int = 800) -> str:
        payload = {"title": title, "content": content, "source": source,
                   "chunk_size": chunk_size}
        op_id = operation_id("knowledge_write", payload)

        def remote_with_mirror():
            result = self.primary.add_document(**payload)
            self.fallback.add_document(**payload)
            return result

        return self.journal.execute("knowledge", op_id, [
            ("elasticsearch", remote_with_mirror),
            ("local", lambda: self.fallback.add_document(**payload)),
        ], cache_result=True)

    def search_knowledge(self, query: str, limit: int = 5) -> list[dict]:
        op_id = operation_id("knowledge_search", {"query": query, "limit": limit})
        return self.journal.execute("knowledge", op_id, [
            ("elasticsearch", lambda: self.primary.search_knowledge(query, limit)),
            ("local", lambda: self.fallback.search_knowledge(query, limit)),
        ])

File: research-workflow-agent-java/src/main/java/com/hyj/researchagent/Resilience.java
package com.hyj.researchagent;

import com.fasterxml.jackson.core.type.TypeReference;

import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.security.MessageDigest;
import java.util.ArrayList;
import java.util.HexFormat;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/** 模型、Embedding、知识库和外部数据源的幂等降级组件。 */
public final class Resilience {
    private Resilience() {}

    @FunctionalInterface public interface Operation<T> { T run() throws Exception; }
    public record Provider<T>(String name, Operation<T> operation) {}

    public static String operationId(String service, Object payload) {
        try {
            String canonical = JsonFiles.JSON.writer().with(com.fasterxml.jackson.databind.SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS)
                    .writeValueAsString(payload);
            String hash = HexFormat.of().formatHex(MessageDigest.getInstance("SHA-256").digest(canonical.getBytes(StandardCharsets.UTF_8)));
            return service + "_" + hash.substring(0,20);
        } catch (Exception error) { throw new IllegalStateException(error); }
    }

    /** 按任务语义和长度确定轻量或强模型，不把路由决策交给模型自身。 */
    public static String classifyComplexity(String text) {
        String lower = text.toLowerCase();
        return text.length()>180 || List.of("比较","综合","评审","架构","多篇","experiment","compare","synthesize","review","architecture")
                .stream().anyMatch(lower::contains) ? "strong" : "light";
    }

    /** 追加切换事件、缓存幂等结果并保存无法完成的外部任务。 */
    public static final class FallbackJournal {
        private final Path events, cache, pending;
        public FallbackJournal(Path dataDir) {
            events=dataDir.resolve("fallback_events.jsonl"); cache=dataDir.resolve("fallback_results.json");
            pending=dataDir.resolve("pending_external_tasks.json");
        }
        public synchronized <T> T execute(String service,String operationId,List<Provider<T>> providers,boolean cacheResult) throws Exception {
            Map<String,Object> stored=JsonFiles.read(cache,new TypeReference<>(){},new LinkedHashMap<>());
            if(stored.containsKey(operationId)) {
                log(operationId,service,"cache","cache","reused","idempotent replay");
                @SuppressWarnings("unchecked") T result=(T)stored.get(operationId); return result;
            }
            List<String> errors=new ArrayList<>();
            for(int i=0;i<providers.size();i++) {
                Provider<T> provider=providers.get(i);
                try {
                    T result=provider.operation().run();
                    log(operationId,service,i==0?provider.name():providers.get(i-1).name(),provider.name(),"succeeded",i==0?"primary":"fallback");
                    if(cacheResult){stored.put(operationId,result);JsonFiles.write(cache,stored);} return result;
                } catch(Exception error) {
                    errors.add(provider.name()+": "+error.getMessage());
                    log(operationId,service,provider.name(),i+1<providers.size()?providers.get(i+1).name():"pending","failed",error.toString());
                }
            }
            throw new IllegalStateException(String.join("; ",errors));
        }
        public synchronized String defer(String service,Map<String,Object> payload,String reason) {
            String id=operationId(service,payload); Map<String,Object> all=JsonFiles.read(pending,new TypeReference<>(){},new LinkedHashMap<>());
            all.putIfAbsent(id,Map.of("id",id,"service",service,"payload",payload,"reason",reason,"status","pending","created_at",JsonFiles.now()));
            JsonFiles.write(pending,all); log(id,service,"all","pending","deferred",reason); return id;
        }
        public void log(String id,String service,String from,String to,String status,String reason) {
            JsonFiles.appendJsonLine(events,Map.of("time",JsonFiles.now(),"operation_id",id,"service",service,
                    "from",from,"to",to,"status",status,"reason",reason));
        }
    }

    @FunctionalInterface public interface ModelClient { ModelResponse create(String model,List<Map<String,Object>> messages,String system,List<Map<String,Object>> tools,int maxTokens) throws Exception; }
    public record ModelResponse(List<Map<String,Object>> content,int inputTokens,int outputTokens) {}
    public record ModelEndpoint(String name,ModelClient client,String model,String tier,boolean primary) {}

    /** 按复杂度排序主备模型，并通过统一接口自动切换。 */
    public static final class ResilientModelClient implements ModelClient {
        private final List<ModelEndpoint> endpoints; private final FallbackJournal journal;
        public ResilientModelClient(List<ModelEndpoint> endpoints,FallbackJournal journal) {
            if(endpoints.isEmpty()) throw new IllegalArgumentException("At least one model endpoint is required");
            this.endpoints=List.copyOf(endpoints);this.journal=journal;
        }
        public ModelResponse createForTask(String complexity,List<Map<String,Object>> messages,String system,List<Map<String,Object>> tools,int maxTokens) throws Exception {
            List<ModelEndpoint> ordered=endpoints.stream().sorted(java.util.Comparator
                    .comparing((ModelEndpoint e)->!e.tier().equals(complexity)).thenComparing(e->!e.primary())).toList();
            String id=operationId("model",Map.of("complexity",complexity,"messages",messages));
            List<Provider<ModelResponse>> providers=ordered.stream().map(endpoint->new Provider<ModelResponse>(endpoint.name(),
                    ()->endpoint.client().create(endpoint.model(),messages,system,tools,maxTokens))).toList();
            return journal.execute("model",id,providers,false);
        }
        public ModelResponse create(String model,List<Map<String,Object>> messages,String system,List<Map<String,Object>> tools,int maxTokens) throws Exception {
            return createForTask("strong",messages,system,tools,maxTokens);
        }
    }

    /** 远程 Embedding 失败时切换本地实现。 */
    public static final class FallbackEmbeddingModel implements KnowledgeService.EmbeddingModel {
        private final KnowledgeService.EmbeddingModel primary,fallback; private final FallbackJournal journal;
        public FallbackEmbeddingModel(KnowledgeService.EmbeddingModel primary,KnowledgeService.EmbeddingModel fallback,FallbackJournal journal){
            this.primary=primary;this.fallback=fallback;this.journal=journal;
        }
        public int dimensions(){return primary.dimensions()>0?primary.dimensions():fallback.dimensions();}
        public List<List<Double>> embedMany(List<String> texts)throws Exception{
            return journal.execute("embedding",operationId("embedding",texts),List.of(
                    new Provider<>("remote",()->primary.embedMany(texts)),new Provider<>("local",()->fallback.embedMany(texts))),false);
        }
    }

    /** Elasticsearch 失败时使用本地知识库，写入结果按 operationId 去重。 */
    public static final class FallbackKnowledgeBase implements KnowledgeService.KnowledgeBase {
        private final KnowledgeService.KnowledgeBase primary,fallback; private final FallbackJournal journal;
        public FallbackKnowledgeBase(KnowledgeService.KnowledgeBase primary,KnowledgeService.KnowledgeBase fallback,FallbackJournal journal){
            this.primary=primary;this.fallback=fallback;this.journal=journal;
        }
        public String addDocument(String title,String content,String source,int chunkSize)throws Exception{
            Map<String,Object> payload=Map.of("title",title,"content",content,"source",source,"chunk_size",chunkSize);
            String id=operationId("knowledge_write",payload);
            return journal.execute("knowledge_write",id,List.of(
                    new Provider<>("elasticsearch",()->{String result=primary.addDocument(title,content,source,chunkSize);fallback.addDocument(title,content,source,chunkSize);return result;}),
                    new Provider<>("local",()->fallback.addDocument(title,content,source,chunkSize))),true);
        }
        public List<Map<String,Object>> search(String query,int limit)throws Exception{
            return journal.execute("knowledge_search",operationId("knowledge_search",Map.of("query",query,"limit",limit)),List.of(
                    new Provider<>("elasticsearch",()->primary.search(query,limit)),new Provider<>("local",()->fallback.search(query,limit))),false);
        }
    }
}
